pkg/telemetry/telemetry.go

```diff
@@ -22,90 +26,168 @@ const actionName = "Piper Library OS"
+		provider = &orchestrator.UnknownOrchestratorConfigProvider{}
 	}
+	t.provider = provider
 
-	if client == nil {
-		client = &piperhttp.Client{}
+	if t.client == nil {
+		t.client = &piperhttp.Client{}
 	}
 
-	client.SetOptions(piperhttp.ClientOptions{MaxRequestDuration: 5 * time.Second, MaxRetries: -1})
+	t.client.SetOptions(piperhttp.ClientOptions{MaxRequestDuration: 5 * time.Second, MaxRetries: -1})
 
+	if t.BaseURL == "" {
+		//SWA baseURL
+		t.BaseURL = "https://webanalytics.cfapps.eu10.hana.ondemand.com"
+	}
+	if t.Endpoint == "" {
+		// SWA endpoint
+		t.Endpoint = "/tracker/log"
+	}
 	if len(LibraryRepository) == 0 {
 		LibraryRepository = "https://github.com/n/a"
 	}
 
-	if len(SiteID) == 0 {
-		SiteID = "827e8025-1e21-ae84-c3a3-3f62b70b0130"
+	if t.SiteID == "" {
+		t.SiteID = "827e8025-1e21-ae84-c3a3-3f62b70b0130"
 	}
 
-	baseData = BaseData{
+	t.baseData = BaseData{
+		Orchestrator:    provider.OrchestratorType(),
+		StageName:       provider.GetStageName(),
 		URL:             LibraryRepository,
 		ActionName:      actionName,
 		EventType:       eventType,
 		StepName:        stepName,
-		SiteID:          SiteID,
-		PipelineURLHash: getPipelineURLHash(), // http://server:port/jenkins/job/foo/
-		BuildURLHash:    getBuildURLHash(),    // http://server:port/jenkins/job/foo/15/
+		SiteID:          t.SiteID,
+		PipelineURLHash: t.getPipelineURLHash(), // http://server:port/jenkins/job/foo/
+		BuildURLHash:    t.getBuildURLHash(),    // http://server:port/jenkins/job/foo/15/
 	}
-	//ToDo: register Logrus Hook
-
+	t.baseMetaData = baseMetaData
 }
 
-func getPipelineURLHash() string {
-	return toSha1OrNA(os.Getenv("JOB_URL"))
+func (t *Telemetry) getPipelineURLHash() string {
+	jobUrl := t.provider.GetJobUrl()
+	return t.toSha1OrNA(jobUrl)
 }
 
-func getBuildURLHash() string {
-	return toSha1OrNA(os.Getenv("BUILD_URL"))
+func (t *Telemetry) getBuildURLHash() string {
+	buildUrl := t.provider.GetBuildUrl()
+	return t.toSha1OrNA(buildUrl)
 }
 
-func toSha1OrNA(input string) string {
+func (t *Telemetry) toSha1OrNA(input string) string {
 	if len(input) == 0 {
 		return "n/a"
 	}
 	return fmt.Sprintf("%x", sha1.Sum([]byte(input)))
 }
 
-// SWA baseURL
-const baseURL = "https://webanalytics.cfapps.eu10.hana.ondemand.com"
-
-// SWA endpoint
-const endpoint = "/tracker/log"
-
-func GetData(customData *CustomData) Data {
-	return Data{
-		BaseData:     baseData,
-		BaseMetaData: baseMetaData,
+// SetData sets the custom telemetry data and base data into the Data object
+func (t *Telemetry) SetData(customData *CustomData) {
+	t.data = Data{
+		BaseData:     t.baseData,
+		BaseMetaData: t.baseMetaData,
 		CustomData:   *customData,
 	}
 }
 
-// Send ...
-func Send(customData *CustomData) {
-	data := GetData(customData)
+// GetData returns telemetryData
+func (t *Telemetry) GetData() Data {
+	return t.data
+}
+
+// Send telemetry information to SWA
+func (t *Telemetry) Send() {
+
+	t.sendCustom()
```

Comment:
why can't this be disabled?

Reply:
it will only be executed if the config provides "customReportingDsn" and we encounter a failure in the step